[17.0][ADD] sale_order_pending_move_menu, sale_order_pending_move_menu_pb, sale_order_pending_move_menu_sds and sale_order_pending_move_menu_wspsa

IriaAlonso opened the pull request:
New menu that show sale line information from stock moves
ticket HT00417

- add commit description --> done
- readme and index --> done
- update translations --> done

Edit: depends on

* https://github.com/solvosci/slv-e-commerce/pull/11

IriaAlonso committed:
[ADD] sale_order_pending_move_menu

New menu that show sale line information from stock moves

Co-Authored-By: David Alonso <david.alonso@solvos.es>

IriaAlonso commented:
<img width="1608" height="253" alt="imagen" src="https://github.com/user-attachments/assets/beff53b3-587b-4c7b-b957-18ba20ce3e9f" />
<img width="1923" height="273" alt="imagen" src="https://github.com/user-attachments/assets/74a5f31c-46ff-46dd-aa9d-553e84b1b259" />
Fields not added --> "Línea", "pedidos restos", "referencia proveedor", "fecha servicio", and product attributes ("color", "tamaño", "marca" and "lote")
@dalonsod can you review it?

dalonsod commented:
@DantePereyra could you review first?

DantePereyra commented:
When entering the form view from the tree view, this is what I see:
<img width="1920" height="556" alt="imagen" src="https://github.com/user-attachments/assets/42e018a6-1c4f-40e9-8b73-4409c437d39e" />

However, in developer mode, this is what appears:
<img width="1915" height="870" alt="imagen" src="https://github.com/user-attachments/assets/aa81139b-c942-4efc-83b7-f1523d4e1bfd" />

Is this the expected behavior?

IriaAlonso commented:
Ignore the `form` view, only `tree` view will be used for the moment.

DantePereyra commented on sale_order_pending_move_menu/views/stock_move_views.xml:
```diff
@@ -0,0 +1,35 @@
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='name']" position="before">
+                <field name="origin" string="Origin"/>
+            </xpath>
```
Maybe added `origin` field as `group by` also.

DantePereyra reviewed (commented):
Overall, it looks good. Minor comment below.
Also, the `origin` field seems like it could be linked to `sale.order`, even though it's originally a `Char`. Perhaps a button next to it could help?
Let’s wait for @dalonsod’s input on this.

dalonsod commented:
> Also, the `origin` field seems like it could be linked to `sale.order`, even though it's originally a `Char`. Perhaps a button next to it could help?

`Origin` field is a `Char` field because depending on move origins it could be have a different content. E.g. if move comes from a purchase order reception, it should be filled with purchase order number, or of move is a consumption line for a MO, with MO number.

So, if this moves surely comes from sales order, this field is expected to be filled with Sales Order number. But there's a more secure and proper way for our original goal: showing here sales order number and, even, a fastlink to it. And this stats using move `sale_line_id` field instead, that let us a definitely more secure path to sales order: https://github.com/odoo/odoo/blob/aeca2b2dcdce255f3fb43d85c79461756db5e924/addons/sale_stock/models/stock.py#L17

IriaAlonso commented:
@dalonsod use of `sale_line_id` instead of `origin` + `group by` added

dalonsod commented on sale_order_pending_move_menu/views/sale_order_views.xml:
```diff
@@ -0,0 +1,27 @@
+        <field name="view_ids" eval="[(5, 0, 0),
+            (0, 0, {'view_mode': 'tree', 'view_id': ref('stock_move_pending_menu_tree')})]"/>
+        <field name="domain">
+        [
+            ('picking_id.state','in',['confirmed','assigned']),
+            ('picking_id.picking_type_code', '=', 'outgoing'),
+            ('origin', '!=', '')
+        ]
```
Changes needed:
* Remove origin condition.
* State condition is incomplete.
* sale link filled required is missing.

dalonsod commented on sale_order_pending_move_menu/views/stock_move_views.xml:
```diff
@@ -0,0 +1,38 @@
+
+    <record id="stock_move_pending_menu_tree" model="ir.ui.view">
+        <field name="name">stock.move.tree (sale_order_pending_move_menu)</field>
+        <field name="model">stock.move</field>
```
Please add lower priority, in order to be only selected by our new menu, and prevent its automatic usage by other menus

dalonsod commented on sale_order_pending_move_menu/views/stock_move_views.xml:
```diff
@@ -0,0 +1,38 @@
+    </record>
+
+    <record model="ir.ui.view" id="stock_move_pending_menu_search">
+        <field name="name">stock.move.search (sale_order_pending_move_menu)</field>
```
Add lower priority management and set it primary. Current search view, already in use, won't be affected.

Once this view is primary and with low priority, you'll need to specify its usage at new action; otherwise it won't be selected there.

dalonsod commented on sale_order_pending_move_menu/views/stock_move_views.xml:
```diff
@@ -0,0 +1,38 @@
+        <field name="name">stock.move.tree (sale_order_pending_move_menu)</field>
+        <field name="model">stock.move</field>
+        <field name="arch" type="xml">
+            <tree>
```
ensure that we cannot edit neither create or delete a move

dalonsod reviewed (changes requested):
See comments

IriaAlonso commented:
view domain update, added lower priorities on views, fix code and not allow to edit, create or delete a move
@dalonsod 

maybe do a hook for `move_sale_order`

dalonsod commented on sale_order_pending_move_menu/views/sale_order_pending_move_menu.xml:
```diff
@@ -0,0 +1,28 @@
+            <field name="search_view_id" ref="stock_move_pending_menu_search"/>
+        <field name="domain">
+        [
+            ('picking_id.state','not in',['done','cancel']),
```
This condition should be actually applied to `move` state, not picking one

dalonsod commented on sale_order_pending_move_menu/views/stock_move_views.xml:
```diff
@@ -0,0 +1,41 @@
+        <field name="model">stock.move</field>
+        <field name="priority">999</field>
+        <field name="arch" type="xml">
+            <tree create="0" delete="0", edit="0">
```
Fix typo:
```suggestion
            <tree create="0" delete="0" edit="0">
```

dalonsod commented:
> maybe do a hook for `move_sale_order`

Go ahead with it

IriaAlonso committed:
[ADD] sale_order_pending_move_menu_pb
Link ‘sale_order_pending_move_menu’ and ‘product_brand’ to add the 'brand' field to the 'pending sales moves' view

dalonsod commented on sale_order_pending_move_menu/models/stock_move.py:
```diff
@@ -0,0 +1,16 @@
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    move_sale_order = fields.Many2one("sale.order", related="sale_line_id.order_id", store=True, string="Order id")
```
Better field naming and label:
```suggestion
    move_sale_order_id = fields.Many2one("sale.order", related="sale_line_id.order_id", store=True, string="Sale Order")
```

dalonsod commented on sale_order_pending_move_menu/models/stock_move.py:
```diff
@@ -0,0 +1,16 @@
+    move_partner_id = fields.Many2one("res.partner", related="picking_id.sale_id.partner_id")
+    move_product_uom_qty_sale = fields.Float(related='sale_line_id.product_uom_qty')
+    move_product_uom_sale = fields.Many2one("uom.uom", related='sale_line_id.product_uom')
+    move_state = fields.Selection(related="picking_id.state")
```
This field should be renamed to `picking_state`, as "Move state" is actually base field `state`.

dalonsod commented on sale_order_pending_move_menu/views/stock_move_views.xml:
```diff
@@ -0,0 +1,41 @@
+                <field name="product_uom" string="Move uom qty"/>
+                <field name="location_id" />
+                <field name="location_dest_id" />
+                <field name="move_state" widget="badge"/>
```
We'll priorize original/native state for move, and preserve related picking state, because we're not sure which state will be the most important for users:
```suggestion
                <field name="state" widget="badge" optional="show"/>
                <field name="picking_state" widget="badge" optional="hide"/>
```

dalonsod commented on sale_order_pending_move_menu/hooks.py:
```diff
@@ -0,0 +1,14 @@
+
+
+def post_init_hook(env):
+    """This hook is used to link a sales order to existing pending moves
```
Once addon is installed and `post_init_hook` is about to run, `move_sale_order_id` is actually filled. And the operation is, again, a python slow one.

We need instead the a `pre_init_hook` that define the field (table column) and pre-fill it before addon is installed, via SQL statement.

dalonsod commented on sale_order_pending_move_menu_link_product_brand/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+        Link ‘sale_order_pending_move_menu’ and ‘product_brand’ to add the 'brand' field to the 'pending sales moves' view
+    """,
+    "author": "Solvos",
+    "license": "LGPL-3",
```
This addon depends on `product_brand`, that belongs to OCA and its license is AGPL-3 (https://github.com/OCA/brand/blob/23bc643bc329d10b33cb951fd48161a0a95ff0a2/product_brand/__manifest__.py#L20).

So this addon has to be licensed `AGPL-3` as well. And add oca_dependencies.txt entry for `OCA/brand`.

dalonsod replied:
This is still missing

> add `oca_dependencies.txt` entry for `OCA/brand`

dalonsod commented on sale_order_pending_move_menu_link_product_brand/models/stock_move.py:
```diff
@@ -0,0 +1,10 @@
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    move_product_brand_id = fields.Many2one("product.brand", related="product_id.product_brand_id", string="Brand", help="")
```
If `help` attribute is not used, remove it

IriaAlonso replied:
I added it to overwrite the `help` attribute for the `product_brand_id` field.

dalonsod reviewed (changes requested):
See comments

IriaAlonso commented:
@dalonsod changes done

dalonsod commented on sale_order_pending_move_menu/hooks.py:
```diff
@@ -0,0 +1,30 @@
+            """
+            ALTER TABLE stock_move
+            ADD COLUMN IF NOT EXISTS move_sale_order_id INTEGER
+            REFERENCES sale_order(id);
```
I don't know how `REFERENCE` actually does. Does it add a foreign key for `sale_order.id`? 

If I'm right, please make the following test. Uninstall the addon and, then, reinstall it again and check if foreign key is later added by Odoo and thus, `REFERENCES` is actually uneeded here.

IriaAlonso commented:
@dalonsod added `oca_dependencies.txt` and deleted `REFERENCES`, it is not need here

dalonsod commented on sale_order_pending_move_menu/hooks.py:
```diff
@@ -0,0 +1,29 @@
+    """
+    if not sql.column_exists(env.cr, "stock_move", "move_sale_order_id"):
+        _logger.info("")
+        env.cr.execute(
+            """
+            ALTER TABLE stock_move
+            ADD COLUMN IF NOT EXISTS move_sale_order_id INTEGER;
+            """
+        )
```
Try `sql.create_column` instead of raw SQL statement, as available tool for Odoo v17, like

https://github.com/solvosci/slv-pos/blob/8cfc4ba57aff8db6e426d23ecdd539010c4ebf36/sale_order_pos_transferred/hooks.py#L12

dalonsod commented on sale_order_pending_move_menu/hooks.py:
```diff
@@ -0,0 +1,29 @@
+    when the sale_order_pending_move_menu module is installed.
+    """
+    if not sql.column_exists(env.cr, "stock_move", "move_sale_order_id"):
+        _logger.info("")
```
Complete logger message, that is currently empty

dalonsod reviewed (changes requested):
See later comments and I'll test it at CmhProd

IriaAlonso commented:
@dalonsod done

dalonsod commented:
Installed at CmhProd, one last thing before requesting LM's review: could you improve state `badge` widgets for both move and picking `state` in `tree` views, like existing views do?

Moves: stock.view_move_tree
Pickings: stock.vpicktree

For those views, state badgets are colored...

IriaAlonso commented:
@dalonsod improved state badges

dalonsod commented:
OK, some decorations are unnecessary (done and cancel moves actually will never be displayed), but we could keep them there.

I'll show this to LM

IriaAlonso committed:
[ADD] sale_order_pending_move_menu_sds
Link ‘sale_order_pending_move_menu’ and sale_delivery_state to add
the 'delivery_status' field to the 'pending sales moves' view

IriaAlonso committed:
[ADD] sale_order_pending_move_menu_wspsa
Link ‘sale_order_pending_move_menu’ and 'website_sale_product_search_advanced' to add
the 'website_partner_ref' field to the 'pending sales moves' view

IriaAlonso commented:
@dalonsod can you review it?

dalonsod commented on oca_dependencies.txt:
```diff
@@ -0,0 +1 @@
+brand
```
Add here reference to missing repos that now are already required due to `sale_delivery_state` and `website_sale_product_search_advanced` dependencies

dalonsod replied:
`website_sale_product_search_advanced` repo is still missing. Please add this (doesn't belong to OCA, and is a work in progress):
```suggestion
brand
slv-e-commerce https://github.com/solvosci/slv-e-commerce 17.0-mig-website_sale_product_search_advanced
```

dalonsod commented on sale_order_pending_move_menu_wspsa/__manifest__.py:
```diff
@@ -0,0 +1,19 @@
+# © 2025 Solvos Consultoría Informática (<http://www.solvos.es>)
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
+{
+    "name": "Sale Order Pending Move Menu Link Website Sale Product Seach Advanced",
```
Installing this addon at CmhProd, once other are installed, raises an error, whose ending text is the following:

`AttributeError: 'NoneType' object has no attribute 'groups'`

dalonsod commented on sale_order_pending_move_menu/models/stock_move.py:
```diff
@@ -0,0 +1,16 @@
+    move_product_uom_sale = fields.Many2one("uom.uom", related='sale_line_id.product_uom')
+    picking_state = fields.Selection(related="picking_id.state")
+    move_warehouse_id = fields.Many2one(related="sale_line_id.warehouse_id")
+    move_user_id = fields.Many2one(related="move_sale_order_id.user_id", store=True)
```
I miss this field in `search` (both search & group by options) and `tree` view

dalonsod reviewed (changes requested):
See comments.

And also, we should simplify addons names, as too long most of them. For this case, we usually use initials for the second addon that is linked (like e.g. https://github.com/solvosci/slv-stock/tree/13.0/stock_picking_mgmt_weight_pot or https://github.com/solvosci/slv-sale/tree/13.0/sale_order_move_menu_spil). Suggested:

* sale_order_pending_move_menu (base one, good name)
* sale_order_pending_move_menu_pb
* sale_order_pending_move_menu_sds
* sale_order_pending_move_menu_wspsa

What do you think?

IriaAlonso commented:
@dalonsod done

dalonsod reviewed (approved):
Let's go...
